L'objectif de ce TP est de se familiariser avec certaines **opérations de base en Text mining (fouille de textes)**.

### 1. Opérations sur des textes bruts

Peut-être encore davantage que d'autres tâches de fouille de données, la fouille de textes et la recherche de documents posent le problème du prétraitement. Les textes et documents sont en effet généralement disponibles sous des formes et des formats divers: en html, xml, doc, pdf, etc. Dans le cadre limité de ce tp, nous supposerons que les documents ont été traités pour être disponibles sous la forme de corpus exploitables, et nous allons nous focaliser sur des opérations d'analyse globale et de ré-description.

Commençons par invoquer nltk et les librairies utiles:

In [1]:
import nltk


Il peut-être utile de lancer la commande : nltk.download() pour charger les ouvrages. Nous chargeons ensuite un texte sur lequel nous ferons diverses opérations.

In [2]:
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

In [4]:
from nltk.book import *

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


Cette commande charge à la fois 9 textes "classiques" et les fonctions prédéfinies de nltk. Le text2 est celui du roman ***Sense and sensibility*** de Jane Austen, écrit en 1811

##### 1.1. Comptages et fréquences

Nous commençons par quelques **opérations élémentaires** : calcul de la *longueur du texte* en mots, *taille du vocabulaire* (nombre de mots différents), *diversité du vocabulaire* (rapport de la longueur du texte sur la taille du vocabulaire).

Calcul du **nombre de mots** de *Sense and Sensibility* :

In [5]:
len(text2)

141576

Calcul de la **taille du vocabulaire** utilisé (pour cela on calcule d'abord l'ensemble des mots par la fonction set)

In [6]:
len(set(text2))

6833

Nous allons maintenant définir une fonction de calcul de la **diversité lexicale**.

In [7]:
def lexical_diversity(text):
    return len(set(text)) / len(text)

Appliquons la fonction à *Sense and Sensibility*.

In [8]:
lexical_diversity(text2)

0.04826383002768831

Appliquons cette même fonction au text1 (*Moby Dick* de Herman Melville en 1851) et au text6 (*Monty Python and the Holy Grail*, film britannique sorti en 1975 et écrit et réalisé par Terry Gilliam et Terry Jones des Monty Python).

In [10]:
print(lexical_diversity(text1))
lexical_diversity(text6)

0.07406285585022564


0.1276595744680851

Il est également possible de compter le nombre de fois où un mot apparaît dans un texte. Par exemple (ici on applique une méthode 'count' à un objet, le texte) : 

In [11]:
text6.count("love")

0